File: Homework 5/wildfire.py
#%% 
import matplotlib.pyplot as plt
import numpy as np 
import matplotlib.colors
import math

omega = [120,120] # dimension of the forest field 
d = 0.5 # density of the forest
n = 200 # number of unit cells
#%% 
N = int(d*n*n) # number of trees 

trees = np.zeros([N,2]) # stores the position of each tree
for i in range(N):
    # uniformly place each tree
    trees[i,:] = [np.random.uniform(0,omega[0]), np.random.uniform(0,omega[1])]

plt.scatter(trees[:,0], trees[:,1], c = 'g', s = 1)
plt.xlim(0, omega[0])
plt.ylim(0, omega[1])
#%% 
# keep track of the states of the trees
unburnt = trees
burning = np.array([[0,0]])

# choose 5 randomly that are burning
rand = np.random.randint(N, size=5)
for i in range(len(rand)):
    tree = unburnt[i,:]
    print(tree)
    burning = np.concatenate((burning, np.array([tree])))
    unburnt = np.delete(unburnt, i, 0)

burning = np.delete(burning, 0, 0)
#%% Test plot forest 
plt.scatter(unburnt[:,0], unburnt[:,1], c = 'g', s = 1)
plt.scatter(burning[:,0], burning[:,1], c = 'r', s = 1)

# %%
# computes the distance with the 2-norm (squared) and initialization
def dist(v, u):
    sum = 0
    for i in range(len(v)):
        sum += (v[i]-u[i])**2
    return sum

#%% Test compute heat map and oxygen map
def update_heatmap_oxygenmap(unburnt, burning):
    temperature = np.zeros(omega)
    oxygen = np.zeros(omega)
    for i in range(temperature.shape[0]):
        for j in range(temperature.shape[1]):
            temp = 0
            oxy = 1
            for k in range(burning.shape[0]):
                temp += 600*np.exp(-1*dist(burning[k],[i,j])/8)
                oxy *= (1 - 0.5*np.exp(-1*dist(burning[k],[i,j]))/32)
            temperature[i,j] = temp
            oxygen[i,j] = oxy

    return [temperature, oxygen]

maps = update_heatmap_oxygenmap(unburnt,burning)
temperature = maps[0]
oxygen = maps[1]
#%% Test Heat Map
heat_map = plt.pcolormesh(temperature)
plt.colorbar(heat_map)
plt.title('Heat Map (degrees Celcius)')
#%% Test Heat Map 
oxygen_map = plt.pcolormesh(oxygen)
plt.colorbar(oxygen_map)
plt.title('Oxygen Map')
#%% Test adding trees 
def update_trees(unburnt, burning,temperature,oxygen):
    new_burning = burning
    new_unburnt = unburnt

    # update burning to burnt
    count = 0
    for i in range(len(burning)):
        if np.random.uniform(0,1) < 0.7:
            new_burning = np.delete(new_burning, i - count, 0)
            count += 1

    # update unburnt to burning
    count = 0
    for i in range(len(unburnt)):
        # get the coordinates of the tree 
        x = math.floor(unburnt[i,0])
        y = math.floor(unburnt[i,1])
        if temperature[x,y] > 300 and oxygen[x,y] > 0.2:
            tree = unburnt[i,:]
            new_burning = np.concatenate((new_burning, np.array([tree])))
            new_unburnt = np.delete(new_unburnt, i - count, 0)
            count += 1

    return [new_unburnt, new_burning]

new_trees = update_trees(unburnt,burning,temperature,oxygen)
unburnt = new_trees[0]
burning = new_trees[1]
#%% Test adding trees
plt.scatter(unburnt[:,0], unburnt[:,1], c = 'g', s = 1)
plt.scatter(burning[:,0], burning[:,1], c = 'r', s = 1)
# %%
time_step = 4
record_times = [1, 2, 3]

fig, ax = plt.subplots(len(record_times),3)

counter = 0
for t in range(time_step):
    print(t)
    new_maps = update_heatmap_oxygenmap(unburnt,burning)
    temperature = new_maps[0]
    oxygen = new_maps[1]
    new_trees = update_trees(unburnt,burning,temperature,oxygen)
    unburnt = new_trees[0]
    burning = new_trees[1]


    if t == record_times[counter]:
        ax[counter,0].scatter(unburnt[:,0], unburnt[:,1], c = 'g', s = 1)
        ax[counter,0].scatter(burning[:,0], burning[:,1], c = 'r', s = 1)
        #ax[counter,0].set_title('Forest Map at time t =' + str(record_times[counter]))


        ax[counter,1].pcolormesh(temperature)
        #ax[counter,1].set_title('Heat Map at time t = ' + str(record_times[counter]))

        ax[counter,2].pcolormesh(oxygen)
        #ax[counter,2].set_title('Oxygen Map at time t = ' + str(record_times[counter]))

        counter += 1

    

    
# %%

File: Homework 5/wildfire.m
d = 0.8; % density
n = 200; % grid dimension 
N = d*n^2; % number of trees

% initialize trees
tree_map = zeros(n,n);
perm = randperm(n^2);
for i = 1:N
    [x,y] = ind2sub([n,n],perm(i));
    tree_map(x,y) = 1;
end

trees = find(tree_map == 1);
perm = randperm(N);
count = 5;
for i = 1:count
    [x,y] = ind2sub([n,n],trees(perm(i)));
    tree_map(x,y) = 2;
end

temp_map = zeros(n,n);
oxy_map = zeros(n,n);

burning = find(tree_map == 2);
for i = 1:n
    for j = 1:n
        temp = 0;
        oxy = 1;
        for k = 1:length(burning)
            [x,y] = ind2sub([n,n],burning(k));
            temp = temp + 600*exp(-1*dist([x,y],[i,j])/8);
            oxy = oxy*(1 - 0.5*exp(-1*dist([x,y],[i,j])/32));
        end
        temp_map(i,j) = temp;
        oxy_map(i,j) = oxy;
    end
end

for t = 1:200
    
    for i = 1:n
        for j = 1:n
            if tree_map(i,j) == 1 && temp_map(i,j) > 300 && oxy_map(i,j) > 0.2
                tree_map(i,j) = 2;
            end 
            if tree_map(i,j) == 2 && rand < 0.7
                tree_map(i,j) = 0;
            end 
        end
    end
    
    figure(1)
    imagesc(tree_map);
    title('Tree Map')
    colormap parula 
    axis('square')
    
    burning = find(tree_map == 2);
    for i = 1:n
        for j = 1:n
            temp = 0;
            oxy = 1;
            for k = 1:length(burning)
                [x,y] = ind2sub([n,n],burning(k));
                temp = temp + 600*exp(-1*dist([x,y],[i,j])/8);
                oxy = oxy*(1 - 0.5*exp(-1*dist([x,y],[i,j])/32));
            end
            temp_map(i,j) = temp;
            oxy_map(i,j) = oxy;
        end
    end
    
    figure(2)
    imagesc(temp_map)
    title('Temperature Map')
    colormap hot
    axis('square')

    figure(3)
    imagesc(oxy_map)
    title('Oxygen Map')
    colormap gray
    axis('square')
    
    if t == 1
        w = waitforbuttonpress;
    end
end

function d = dist(u,v)
    d = (u(1)-v(1))^2 + (u(2)-v(2))^2;
end
